design/ceph/osd-migration.md

```diff
@@ -0,0 +1,106 @@
+  -
+
+- Example:
+```yaml
+storage:
+  osdStore:
+    type: bluestore_rdr
+    updateStore: yes-really-update-store
+```
+
+- New deployments will use this `osdStore.type` while creating the OSDs.
+- If the user wants to change the `osdStore.type` field for an existing cluster, then they need to provide `spec.Storage.osdStore.updateStore:yes-really-update-store` in order to migrate the existing osds to use update osdStore.
+
+- Add a new label `osd-store:<osd store type>` to all the OSD pods.
+- This label will help to identify the current backend store being used for the OSD and enable the operator to decide if the
+  OSD should be migrated incase the user changes the backend store type in the spec (`spec.Storage.osdStore.type`).
+
+
+### NEW CLUSTERS
+- Currently the osd backend store is hardcoded to `bluestore` throughout the operator code.
+- Add a new environment variable called `ROOK_OSD_STORE_TYPE` in the OSD prepare and OSD pods. This will contain the OSD store type. Use this environment variable to prepare and activate the cluster.
+
+#### OSD Prepare
+- `ceph-volume prepare` command should be updated to use backend store provided in the `spec.storage.osdStore`. If no value is provided for `spec.storage.osdStore` then default to `bluestore` backend.
+
+  - **OSD on disks**:
+    - RAW MODE
+      - `ceph-volume lvm prepare --bluestore_rdr ...`
+    - LWM MODE
+      - `ceph-volume raw prepare --bluestore_rdr ...`
+  - **OSD on PVCs**:
+    -
+
+#### OSD Create
+- `OSD_STORE_FLAG` should use the value provided in the `spec.storage.osdStore`. If no values are provided by the user, then `OSD_STORE_FLAG` should default to `bluestore`
+- `ceph-volume` commands should use the `OSD_STORE_FLAG` while activating the OSDs.
+  - **OSD on disks**:
+    - RAW MODE
+    `ceph-volume raw activate --device "$DEVICE" --no-systemd --no-tmpfs`
+
+    - LVM MODE
+    `ceph-volume lvm activate --no-systemd "$OSD_STORE_FLAG" "$OSD_ID" "$OSD_UUID"`
+
+  - **OSD on PVCs**:
+
+
+
+### EXISTING CLUSTERS
+- After upgrading an existing rook ceph cluster to a version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+  - Example:
+```yaml
+storage:
+  osdStore:
+    type: <desired-backend-store>
+    updateStore: yes-really-update-store
```

Comment:
I still wonder if we need specific flags that indicate the conversion intended. For example, let's say they have these settings:
```yaml
    type: bluestore-rdr
    updateStore: yes-really-update-store
```
Then they update this in the future when seastore exists:
```yaml
    type: seastore
    updateStore: yes-really-update-store
```
We would upgrade to seastore automatically. This does seem acceptable, but also is not as safe as if we had used a specific flag. With a specific flag, if they had the following, no upgrade would proceed since it doesn't indicate to update to seastore:
```yaml
    type: seastore
    updateStore: yes-really-update-to-bluestore-rdr
```
So I feel we do need the specific flag.

Reply:
Lets discuss more about this in the huddle.

Reply:
Did we discuss? I don't recall a resolution, but maybe because I was out so much last month. :)

Reply:
We did but there was no conclusion.

Reply:
I'm ok leaving it as `yes-really-update-store` instead of something more specific, since we're so careful during the replacement of OSDs one-by-one.